### PR TITLE
Request to deprecate `ExtensionRegistry->load`

### PR DESCRIPTION
**Pull request by hexmode**
### Setup and configuration

- MW version: 1.33/1.34?

### Issue

https://gerrit.wikimedia.org/r/#/c/mediawiki/core/+/504032/

https://github.com/SemanticMediaWiki/SemanticGlossary/blob/02ed40ccd4b6fb8f8db50e1477092fb8b2d7c53f/SemanticGlossary.php#L25-L30

### CODE REVIEW

**Review by akuckartz — APPROVED** [review]

**Comment by hexmode** [comment]
I'm not sure how to address the failures.  This one is confusing:
```
1) SG\Tests\Integration\MwDBSQLStoreIntegrationTest::testPageCreateDeleteStoreIntegration

TypeError: Return value of SG\Tests\Integration\MwDBSQLStoreIntegrationTest::setUp() must be an instance of SG\Tests\Integration\void, none returned
```

**Comment by hexmode** [comment]
Ah... That one is saying [it needs 7.1](https://ourcodeworld.com/articles/read/942/how-to-solve-composer-install-update-error-return-value-of-packageversionsinstalleractivate-must-be-an-instance-of-packageversionsvoid-none-returned).

7.2 is confusing, though.
```
PHP Fatal error:  Declaration of Mock_DIWikiPage_c1bf0d64::isSubEntityOf(string $prefix) must be compatible with SMW\DIWikiPage::isSubEntityOf(string $prefix): bool in /home/travis/build/SemanticMediaWiki/mw/vendor/phpunit/phpunit-mock-objects/src/Framework/MockObject/Generator.php(290) : eval()'d code on line 1
```

I imagine that is a problem with phpunit or php-mocks, but I don't know what to do.

**Comment by hexmode** [comment]
pinging @JeroenDeDauw in case he has a clue.

**Comment by JeroenDeDauw** [comment]
Error hints at missing boolean return type.

Perhaps the issue is too old PHPUnit for newer PHP version? Seen some of that in various places in recent history. Can just upgrade PHPUnit in that case

**Comment by hexmode** [comment]
> Perhaps the issue is too old PHPUnit for newer PHP version?

Looks like [it needs phpunit 5+](https://github.com/sebastianbergmann/phpunit-mock-objects/issues/283)

File: composer.json [hunk 1]
```diff
@@ -37,6 +37,9 @@
 		"mediawiki/semantic-media-wiki": "^2.4|^3.0",
 		"mediawiki/lingo": "^2.0.3|^3.0"
 	},
+	"require-dev": {
+		"phpunit/phpunit": "^6.5"
+	},
```
Comment (JeroenDeDauw) [inline composer.json:42]: Not sure what this is good for

**Review by JeroenDeDauw — APPROVED** [review]

**Comment by kghbln** [comment]
Since we got relevant approval I just merged. A field test showed that the fatal for MW 1.35 is resolved. However I am not sure if the extension is still working as expected. This may be a separate issue.

**Comment by hexmode** [comment]
> I found this issue following the [extension talks](https://www.mediawiki.org/wiki/Extension_talk:Semantic_Glossary) where SemanticGlossary was labeled "Broken" with 1.34. Also googling the above mentioned error led me here, I'm producing it with a rather fresh installation of MediaWiki 1.35 and SemanticMediaWiki 3.2.2:
> 
> `Uncaught Error: Call to undefined method ExtensionRegistry::load() in SemanticGlossary`

Did you try using the latest from git (e.g. "dev-master" in composer.local.json)?

**Comment by Hannes-Ole** [comment]
Thank you for answering. I tried, but that didn't fix it.
I tried via
`git clone https://github.com/SemanticMediaWiki/SemanticGlossary.git`
This resulted in error (Semantic Glossary requires Lingo to be installed.)
Manual install of Lingo (via composer.local.json):
		```"mediawiki/lingo": "~3.0",```
didn't resolve this until adding `#wfLoadExtension( 'Lingo' );` to LocalSettings.php (against SemanticGlossary recommendation) That removed the dependency error and leaves the wiki functional, but with a php-error displayed on top:
```
Warning: Invalid argument supplied for foreach() in /var/www/webroot/ROOT/extensions/Lingo/src/Lingo.php on line 76
```

Then deinstalled both (Lingo via composer and SemanticGlossary manually) and added:
		```"mediawiki/semantic-glossary": "dev-master"```
to composer.local.json and ran `composer update`
again resulted in dependency error (Semantic Glossary requires Lingo to be installed.)
Again installed Lingo manually with
		```"mediawiki/lingo": "dev-master",```
Dependency error persisted until adding `#wfLoadExtension( 'Lingo' );` to LocalSettings.php

Again this removes the dependency error and leaves the wiki functional, but with a similar php-error displayed on top (this time line 89 instead of 76):
```
Warning: Invalid argument supplied for foreach() in /var/www/webroot/ROOT/extensions/Lingo/src/Lingo.php on line 89
```

The dependency error displayed when Lingo is NOT manually installed/loaded reads:
```
Semantic Glossary requires Lingo to be installed.
Version
Semantic MediaWiki:
3.2.2 (smw:2020-04-18)
MediaWiki:
1.35.0
Code:
ERROR_EXTENSION_DEPENDENCY
```
and gives advice about adding `enableSemantics` in LocalSettings.php, but this is already done (before adding `wfLoadExtension( 'SemanticGlossary' );`)

**Comment by kghbln** [comment]
In the meantime a new issue was created https://github.com/SemanticMediaWiki/SemanticGlossary/issues/56

**Comment by Hannes-Ole** [comment]
Thanks for pointing it out. So MW 1.35+ is "officially" not supported yet.
Moving to "dev-master" removes the error `Uncaught Error: Call to undefined method ExtensionRegistry::load() in SemanticGlossary` but requires manual installation and enabling of Lingo. Also there seems to be a problem with the Lingo code.
The php error is only display on the "Special:Version" page... is it safe to use SemanticGlossary in the combination:
```
		"mediawiki/lingo": "dev-master",
		"mediawiki/semantic-glossary": "dev-master"
```
or should I rather wait until 1.35+ is supported?

**Comment by kghbln** [comment]
> So MW 1.35+ is "officially" not supported yet.

That's a good way of putting it. I hope that a withe night shows up to make this happen to allow users of MW 1.31 an upgrade path. Until then it does not make sense to have it on MW 1.35

I cannot remember why I de-installed this extension on [sandbox](https://sandbox.semantic-mediawiki.org/wiki/Main_Page) for a second time however I once got it working on MW 1.35 but it did not show tooltips at all.

**Pull request merged**